<a href="https://colab.research.google.com/github/radhikasethi2011/btechproj/blob/main/Colab/data_generation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
!git clone https://github.com/radhikasethi2011/btechproj.git

fatal: destination path 'btechproj' already exists and is not an empty directory.


left

In [46]:

import pandas as pd
import numpy as np
from random import seed
df = pd.read_csv('/content/btechproj/Data/fruitfly/Fruitfly Datasets/shift_200/left.txt', delimiter="\t")
df

FoldID                 EventId  start_index  \
0         A         chr2L:7316-7516           -5   
1         A       chr2L:21243-21443           -5   
2         A       chr2L:66022-66222           -5   
3         A       chr2L:71092-71292           -5   
4         A       chr2L:85916-86116           -5   
...     ...                     ...          ...   
4118      A  chrX:21883414-21883614           -5   
4119      A  chrX:21890084-21890284           -5   
4120      A  chrX:21930715-21930915           -5   
4121      A  chrX:21943991-21944191           -5   
4122      A  chrX:22226296-22226496           -5   

                                                    seq  Bound  
0     TAGAGAGGAGAGGACAATATTATAATTGTAGACCGTTTTAAACACT...      0  
1     CACAGATAATAATACGACCGGTCAGAGCTTAACCGTGTATCTGTTT...      0  
2     TTTAGAGTAAAAGTTACCCCAAGATATTGATTCTTATGTATAGATT...      0  
3     TATTTAAAGAATAACACGATTTGGTTGAAAAAATTTCCTTACGTAT...      0  
4     CACTGGGAAGCGTTCCCTGTGTGGGAAGCCTCGTCTTATCATAACA...      0  
...                                                 ...    ...  
4118  CCGTATTTTTATAAAGGAAAGGGGCTAGGAATAAAGTTAGGGGTAA...      0  
4119  CTTCCAGCGAAGAGATAACCAAACAACAGTCCAAAGAGTAATGAGC...      0  
4120  CGCAAGCGCAAGGCGAGCAATGAATGCTGCTCGGCATTTCCACTGC...      0  
4121  CTGGCTATGCTTTGACTTGCACAGTCCCACTTACCTATGTACATGC...      0  
4122  TTATCACAAAAAATCGGGGGTTTCTATTAAAAATTTAAGTGGTCTG...      0  

[4123 rows x 5 columns]

In [47]:
df1 = pd.read_csv('/content/btechproj/Data/fruitfly/Fruitfly Datasets/shift_200/tad.txt', delimiter="\t")
df1

FoldID                 EventId  start_index  \
0         A       chr2L:12482-12682           -5   
1         A       chr2L:43932-44132           -5   
2         A       chr2L:68857-69057           -5   
3         A       chr2L:78079-78279           -5   
4         A       chr2L:90108-90308           -5   
...     ...                     ...          ...   
4118      A  chrX:21887049-21887249           -5   
4119      A  chrX:21910699-21910899           -5   
4120      A  chrX:21936579-21936779           -5   
4121      A  chrX:22085309-22085509           -5   
4122      A  chrX:22237079-22237279           -5   

                                                    seq  Bound  
0     GCTACAACCGTGCATACTCTGCAGTGTTTCCGGGGATATGCGACAA...      1  
1     AAAAGGGTTACAACAAATTGAAAGTCGAGTAACCATAAATTTTTGG...      1  
2     CGTGGAGGAAGCACTGCCAGAGCACATGATCTTTGCGGCGCCAAAA...      1  
3     GTCCTGAATTGACGTATATGTCTACATGTGTTTCCAAAGACAGTTG...      1  
4     AAATAAAAACAAATATTTACGAAAATATGTTCAGATCCGAATTTGT...      1  
...                                                 ...    ...  
4118  TTCCTCTTCCCCAACACAAACAGCGCAAGAAGCCCTAGCAATTCGG...      1  
4119  ATATTTTTATAGGCTTACTAATGCTAGGCAGTGGGAGTTATCCACC...      1  
4120  CACGATCATCAGTTCCCAACAACTTGTTTCGACTCTTCGAAGACAA...      1  
4121  TTCGCTGCGCCAAGACAACACATGCTCAATTTCGGTAATGCACCGT...      1  
4122  TAAGTCGCCCGATGGAAAACAGCTTTTAGCCTAATCACAATATTAT...      1  

[4123 rows x 5 columns]

for markov 

In [48]:
from sklearn.model_selection import train_test_split
train_left, test_left = train_test_split(df, random_state = 42, test_size = 0.20)
train_tad, test_tad = train_test_split(df1, random_state = 42, test_size = 0.20)

In [49]:
train_left = train_left.sample(frac=1).reset_index(drop=True)
test_left = test_left.sample(frac=1).reset_index(drop=True)
train_tad = train_tad.sample(frac=1).reset_index(drop=True)
test_tad = test_tad.sample(frac=1).reset_index(drop=True)

In [50]:
print('len(train_left', len(train_left))
print('len(train_tad', len(train_tad))
print('len(test_left', len(test_left))
print('len(test_tad', len(test_tad))

len(train_left 3298
len(train_tad 3298
len(test_left 825
len(test_tad 825


In [51]:
type(train_left)

pandas.core.frame.DataFrame

In [52]:
test_markov = pd.concat([test_left, test_tad], axis=0).sample(frac=1).reset_index(drop=True)

In [53]:
test_markov

FoldID                  EventId  start_index  \
0         A     chrX:9552363-9552563           -5   
1         A  chr2L:13011643-13011843           -5   
2         A     chrX:5953883-5954083           -5   
3         A  chr2R:16184051-16184251           -5   
4         A    chr3R:8336180-8336380           -5   
...     ...                      ...          ...   
1645      A    chr2R:8400987-8401187           -5   
1646      A  chr3R:27637620-27637820           -5   
1647      A    chr3R:6609458-6609658           -5   
1648      A    chr3L:8322086-8322286           -5   
1649      A   chrX:16685765-16685965           -5   

                                                    seq  Bound  
0     AAATACACACACACACACACACGTAACACGCACAAGCAGGCGCTCT...      0  
1     CCTGGCCAGGAGCTCATCCGGATTAGTATCATTAAAACCAACCTTC...      1  
2     TACAATTAATATAATATTGTTCTCCTGTGCTGCAGCTGCTTGTGAA...      1  
3     TTAAAAACCAACAGCGAAAACAATAACAAGCCCAACTACAAGATAG...      1  
4     ACCAGAGCGGCAAGATCCAGGACGTGCTGCTGGTGGACGTGGCCCC...      0  
...                                                 ...    ...  
1645  TCCACCTCCTCTGTAGCGCCAACGACCTGGAGTCAAGCAATTTAGT...      1  
1646  GATCTCCTTGTACTTCTTCTGTACCTCGGGCGTCCTGCGCTGGGCG...      0  
1647  GAACGGGTTTTTCGCGTGCGCCAACAATGTCCACTCTGTAGGTAAA...      0  
1648  TTCGGATGACAATAAATCCCGATCAATTAAAAATCGAAGTGGAAAA...      0  
1649  TATAGTTTTTATGTTCCAAAACGCAACATATAAAGAAGAGTATTTT...      0  

[1650 rows x 5 columns]

In [54]:
train_left.to_csv('train_left.txt', index=None, sep='\t')
train_tad.to_csv('train_tad.txt', index=None, sep='\t')
test_markov.to_csv('test_markov.txt', index=None, sep='\t')

for DL 

left

In [106]:
left = np.empty((3298,200,4), dtype=float)
left.shape

(3298, 200, 4)

In [107]:
train_left

FoldID                  EventId  start_index  \
0         A    chr3R:1604917-1605117           -5   
1         A  chr2L:11498653-11498853           -5   
2         A  chr3R:24162252-24162452           -5   
3         A      chr3R:254323-254523           -5   
4         A   chrX:11914275-11914475           -5   
...     ...                      ...          ...   
3293      A  chr3L:19508833-19509033           -5   
3294      A  chr3R:22691859-22692059           -5   
3295      A  chr2R:17953449-17953649           -5   
3296      A    chr2R:7728385-7728585           -5   
3297      A    chr3R:8500006-8500206           -5   

                                                    seq  Bound  
0     TTCTAAAACTTAGCGGAATTTCCCCTCTCTTTCTAACGGTTATTTG...      0  
1     CGTCTGCTCCTTCTTCAGCTGCTCCACCTTGCAGGTACCCAGTTTG...      0  
2     TAATAGAGATGTCCCAACCGGCTCGCAACTGATTTAATTTATTTAA...      0  
3     TCTGCATCACCTACACATGGACCTGGAACACAAAACACAATAGTTT...      0  
4     CAACGCAATGGCTCAAACAACGATTTAGTTGAATTAATGTTCAGAA...      0  
...                                                 ...    ...  
3293  AAATGAAGATGAGAACAAGGGCAGCTTCGTAGTCGATATACCCGAT...      0  
3294  GACCGTGCTTCTGCATGAACTGTTTGATGTCCTCCAGCTTGGCGTT...      0  
3295  GTTGTTTGTGCAAAAGTCGACTTCCGCATATGGTCCACCAATCGTT...      0  
3296  TAGTACCAATCCCTTGGCGAGTCTGGTACCTGATGGTAAATAAACG...      0  
3297  TGTAAGCTGACATGCTTAACTATTTGCTTCGTCATTCTATTTTGAC...      0  

[3298 rows x 5 columns]

In [108]:
def onehote(sequence):
    mapping = {"A": 0, "C": 1, "G": 2, "T": 3}
    seq2 = [mapping[i] for i in sequence]
    return np.eye(4)[seq2]

for i in range(len(train_left)):
  left[i] = onehote(train_left['seq'][i])

In [109]:
import regex as re
with open('left_matrix.txt', 'w') as fout:
    fout.write(u'#'+'\t'.join(str(e) for e in left.shape)+'\n')
    left.tofile(fout)

with open('left_matrix.txt', 'rb') as f:
    line=f.readline().decode('ascii')
    if line.startswith('#'):
        shape=tuple(map(int, re.findall(r'(\d+)', line)))
    else:
        raise IOError('Failed to find shape in file')    

    result2=np.fromfile(f)
    result3=result2.reshape(shape)

In [110]:
result3.shape

(3298, 200, 4)

    X_train = data["train_data"]

    Y_train = data["train_labels"]

    X_validation = data["val_data"]

    Y_validation = data["val_labels"]

    X_test = data["test_data"]

    Y_test = data["test_labels"]

tad

In [111]:
tad = np.empty((3298,200,4), dtype=float)
tad.shape

(3298, 200, 4)

In [112]:

for i in range(len(train_tad)):
  tad[i] = onehote(train_tad['seq'][i])

In [113]:
with open('tad_matrix.txt', 'w') as fout:
    fout.write(u'#'+'\t'.join(str(e) for e in tad.shape)+'\n')
    tad.tofile(fout)

with open('tad_matrix.txt', 'rb') as f:
    line=f.readline().decode('ascii')
    if line.startswith('#'):
        shape=tuple(map(int, re.findall(r'(\d+)', line)))
    else:
        raise IOError('Failed to find shape in file')    

    result22=np.fromfile(f)
    result33=result22.reshape(shape)

In [114]:
result33.shape

(3298, 200, 4)

testing data

In [133]:
testt = np.empty((1650,200,4), dtype=float)
testt.shape

(1650, 200, 4)

In [134]:
def onehote(sequence):
    mapping = {"A": 0, "C": 1, "G": 2, "T": 3}
    seq2 = [mapping[i] for i in sequence]
    return np.eye(4)[seq2]

for i in range(len(test_markov)):
  testt[i] = onehote(test_markov['seq'][i])

In [135]:
import regex as re
with open('test_matrix.txt', 'w') as fout:
    fout.write(u'#'+'\t'.join(str(e) for e in testt.shape)+'\n')
    testt.tofile(fout)

with open('test_matrix.txt', 'rb') as f:
    line=f.readline().decode('ascii')
    if line.startswith('#'):
        shape=tuple(map(int, re.findall(r'(\d+)', line)))
    else:
        raise IOError('Failed to find shape in file')    

    result222=np.fromfile(f)
    result333=result222.reshape(shape)

In [136]:
result333.shape

(1650, 200, 4)

concatinate left and tad training 

In [119]:
merge = np.empty((6596,200,4), dtype=float)
merge = np.concatenate([left, tad], 0)
merge.shape

(6596, 200, 4)

merge label matrices - training

In [120]:
from numpy import array
from numpy import argmax

from tensorflow.keras.utils import to_categorical
# define example
data = np.concatenate([train_left['Bound'],train_tad['Bound']], 0) #left , tad 
data = array(data)
print(data)
# one hot encode
encoded1 = to_categorical(data)
print(encoded1)

[0 0 0 ... 1 1 1]
[[1. 0.]
 [1. 0.]
 [1. 0.]
 ...
 [0. 1.]
 [0. 1.]
 [0. 1.]]


In [121]:
encoded1.shape

(6596, 2)

In [122]:
df2 = pd.DataFrame()
df2['data'] = merge.tolist()
df2['label'] = encoded1.tolist()

In [123]:
df2 = df2.sample(frac = 1).reset_index(drop=True)
df2

data       label
0     [[0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  [1.0, 0.0]
1     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...  [1.0, 0.0]
2     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...  [0.0, 1.0]
3     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...  [1.0, 0.0]
4     [[0.0, 0.0, 0.0, 1.0], [0.0, 1.0, 0.0, 0.0], [...  [1.0, 0.0]
...                                                 ...         ...
6591  [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...  [1.0, 0.0]
6592  [[0.0, 0.0, 1.0, 0.0], [0.0, 1.0, 0.0, 0.0], [...  [0.0, 1.0]
6593  [[1.0, 0.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [...  [0.0, 1.0]
6594  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...  [1.0, 0.0]
6595  [[0.0, 1.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [...  [0.0, 1.0]

[6596 rows x 2 columns]

merge label matrices - testing 

In [125]:
from numpy import array
from numpy import argmax

from tensorflow.keras.utils import to_categorical
# define example
data = test_markov['Bound'] 
data = array(data)
print(data)
# one hot encode
encoded11 = to_categorical(data)
print(encoded11)
print(encoded11.shape)

[0 1 1 ... 0 0 0]
[[1. 0.]
 [0. 1.]
 [0. 1.]
 ...
 [1. 0.]
 [1. 0.]
 [1. 0.]]
(1650, 2)


In [137]:
test = pd.DataFrame()
test['data'] = testt.tolist()
test['label'] = encoded11.tolist()

In [138]:
test

data       label
0     [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  [1.0, 0.0]
1     [[0.0, 1.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [...  [0.0, 1.0]
2     [[0.0, 0.0, 0.0, 1.0], [1.0, 0.0, 0.0, 0.0], [...  [0.0, 1.0]
3     [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...  [0.0, 1.0]
4     [[1.0, 0.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [...  [1.0, 0.0]
...                                                 ...         ...
1645  [[0.0, 0.0, 0.0, 1.0], [0.0, 1.0, 0.0, 0.0], [...  [0.0, 1.0]
1646  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  [1.0, 0.0]
1647  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  [1.0, 0.0]
1648  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...  [1.0, 0.0]
1649  [[0.0, 0.0, 0.0, 1.0], [1.0, 0.0, 0.0, 0.0], [...  [1.0, 0.0]

[1650 rows x 2 columns]

In [129]:
train, val = train_test_split(df2, random_state = 42, test_size = 0.25)

In [130]:
print('len(train', len(train))
print('len(val', len(val))


len(train 4947
len(val 1649


final splits: train 4947, val 1649, test 1650 - DL 

train 6596, test 1650 - MARKOV

In [139]:
train_data = np.array(train.data.tolist())
train_label = np.array(train.label.tolist())
test_data = np.array(test.data.tolist())
test_label = np.array(test.label.tolist())
val_data = np.array(val.data.tolist())
val_label = np.array(val.label.tolist())

In [140]:
print("train_data.shape: ",train_data.shape)
print("train_label.shape: ",train_label.shape)
print("test_data.shape: ",test_data.shape)
print("test_label.shape: ",test_label.shape)
print("val_data.shape: ",val_data.shape)
print("val_label.shape: ",val_label.shape)

train_data.shape:  (4947, 200, 4)
train_label.shape:  (4947, 2)
test_data.shape:  (1650, 200, 4)
test_label.shape:  (1650, 2)
val_data.shape:  (1649, 200, 4)
val_label.shape:  (1649, 2)


In [141]:
import h5py
with h5py.File('training_200seq_2class.hdf5', 'w') as f:
  f.create_dataset('train_data', data=train_data)
  f.create_dataset('train_labels', data=train_label)
  f.create_dataset('test_data', data=test_data)
  f.create_dataset('test_labels', data=test_label)
  f.create_dataset('val_data', data=val_data)
  f.create_dataset('val_labels', data=val_label)

In [142]:
!gzip /content/training_200seq_2class.hdf5